### ISSUE TITLE
New `new_relic_logs` sink

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#1261]
@Jeffail has more insight into what's required here, but from the sound of it, it appears this will simply wrap the `http` sink. If that's the case, I'd like to think a little more about how these types of sinks are created, because we'll be creating quite a few more (ex: #1000).

@Jeffail if there are any pending issues or questions feel free to add them here and we can weigh in.

**Comment by Jeffail** [comment]
According to the [docs](https://docs.newrelic.com/docs/logs/new-relic-logs/log-api/introduction-log-api) there are two formats for sending data to New Relic over HTTP, the first is a simple JSON object:

```json
POST /log/v1 HTTP/1.1
Host: log-api.newrelic.com
Content-Type: application/json
X-Insert-Key: <YOUR_INSIGHTS_INSERT_KEY>
Accept: */*
Content-Length: 133
{
     "timestamp": <TIMESTAMP-OF-EVENT>,
     "message": "User 'xyz' logged in",
     "service": "login-service",
     "hostname": "login.example.com"
}
```

The second is array based, with the ability to send >1 logs in a single request and the ability to set common fields:

```json
POST /log/v1 HTTP/1.1
Host: log-api.newrelic.com
Content-Type: application/json
X-Insert-Key: <YOUR_INSIGHTS_INSERT_KEY>
Accept: */*
Content-Length: 319
[{
   "common": {
     "attributes": {
       "service": "login-service",
       "hostname": "login.example.com"
     }
   },
   "logs": [{
       "timestamp": <TIMESTAMP-OF-EVENT>,
       "message": "User 'xyz' logged in"
     },{
       "timestamp": <TIMESTAMP-OF-EVENT>,
       "message": "User 'xyz' logged out",
       "attributes": {
         "auditId": 123
       }
     }]
}]
```

The first version doesn't support newline delimited JSON, the second version would require a custom sink implementation to build that format.

However, it looks as though the simplified API also allows you to post JSON arrays e.g. `[{"message":"foo"},{"message":"foo2"}]`, so now that the `http` sink supports the `json` encoding option that means we can support New Relic with this config:

```toml
[sinks.nr]
inputs = ['foo']
type = 'http'
uri = 'https://log-api.newrelic.com/log/v1'
encoding = 'json'

[sinks.nr.headers]
Content-Type = "application/json"
X-License-Key = "xxxxx"
```

So I think the logical next step is to implement a `type = "new_relic"` sink with a config format like:

```toml
[sinks.nr]
inputs = ['foo']
type = 'new_relic'
license_key = 'xxxxx' # optional
insert_key = 'xxxxx' # optional
region = 'us' # enum, options are 'us' or 'eu', default 'us'
```

Which under the hood simply creates the equivalent `http` config and spawns it.

## THINGS LEFT TO CONSIDER / DECISIONS TO BE MADE

- We're currently working on exposing metrics for components within Vector, if we're opaquely creating an `http` sink internally then might this cause confusion for the user?

**Comment by Jeffail** [comment]
Implementation (https://github.com/timberio/vector/commit/196e025f4762ab9ccde6cf7bed868457aff92d7d) is very straight forward. The only potential issue I see with maintaining this and other wrappers in future would be keeping up with optional fields as they are added to the `http` sink, but as long as the sinks continue to work I don't see that being a problem for users.

**Comment by binarylogic** [comment]
This looks great. A few things we should address:

- [ ] Rate limiting. We need to adjust the defaults to those that align with New Relic's service. This includes all `batch_*`, `request_*`, and `rate_*` options.
- [ ] I assume the identifier in the logs will reference the `new_relic_log` sink and not `http`, correct?
- [ ] Can we get a test or 2 to ensure the sink does not regress?
- [ ] Does new relic have any special fields in the schema we should consider? For example, if they prefer `hostname` instead of `host` it might be worth addressing that here. The same goes for `message`, `timestamp`, and so on.

**Comment by Jeffail** [comment]
1. Yes agree 100% about rate limiting, it's easy to implement.
2. Not currently, I imagine there's possibly a clean way to change the logs and metrics but it needs investigating.
3. Agree.
4. We align on `timestamp` and `message`, everything else is documented as an "extra" field. However, in practice their logs UI also shows `hostname` and `service_name` along with those two. Our `host` key appears when you select a log along with all other fields but doesn't populate the `hostname` field in its absense.

I think at the very least we should add a transform example to the docs that shows how to populate those two fields (copy `host` to `hostname`, set `service_name` from env var). I'm not sure it's worth engineering our way into populating those fields by default as I think it would ultimately mean forking the http sink.

**Comment by binarylogic** [comment]
> I think at the very least we should add a transform example to the docs that shows how to populate those two fields (copy host to hostname, set service_name from env var)

Sure, this is a fine start. Although, if there is meaningful value to mapping these fields then we should do it in the future. I think it is within the scope of a sink to do anything that maximizes the UX with the downstream service. We've discussed being able to chain/couple components to create a larger component. Maybe that would solve this when we get to it.

**Comment by Jeffail** [comment]
Turns out logging (and metrics) is fine as we use the `sink_type()`.

I played around with gzip compression and couldn't get it to work. I suspect it's due to how we construct our headers but needs further investigation.

I also looked into rate limiting/batching and there are documented usage limits: https://docs.newrelic.com/docs/licenses/license-information/general-usage-licenses/new-relic-data-usage-limits-policies, but they don't really help us here. I've reached out so hopefully we'll either get some guidance or we can plan out throughput tests to find some sensible defaults.
